### PR TITLE
Improve docs for `typing.TypeAlias`

### PR DESCRIPTION
**Pull request by AlexWaygood**
The current docs for `typing.TypeAlias` are pretty bare on what the advantages are for doing this:

```py
X: TypeAlias = list[int]
```

Compared to this:

```py
X = list[int]
```

I've added a motivating example in this PR.

While `typing.TypeAlias` is now deprecated following PEP-695, I think it's still worth improving the docs here. Lots of users won't be able to use the new syntax for a while, since it can't be backported to Python <3.12, and many users will want to continue supporting older Python versions for a while to come. Moreover, the fact that `typing.TypeAliasType` (the new mechanism introduce by PEP-695) has such a similar name is pretty confusing, in my opinion. It reinforces the need for us to have good documentation for both the new and the old symbol.

----
:books: Documentation preview :books:: https://cpython-previews--105372.org.readthedocs.build/

### CODE REVIEW

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -805,19 +805,41 @@ These can be used as types in annotations and do not support ``[]``.
...
+
+   .. testcode::
+
+      from typing import Generic, TypeAlias, TypeVar
+
+      T = TypeVar("T")
+
+      # "Container" does not exist yet,
+      # so we have to use quotes for the forward reference on Python <3.12.
+      # Using `TypeAlias` tells the type checker that this is a type alias declaration,
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
      # Using ``TypeAlias`` tells the type checker that this is a type alias declaration,
```

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -805,19 +805,41 @@ These can be used as types in annotations and do not support ``[]``.
...
+      T = TypeVar("T")
+
+      # "Container" does not exist yet,
+      # so we have to use quotes for the forward reference on Python <3.12.
+      # Using ``TypeAlias`` tells the type checker that this is a type alias declaration,
+      # not a variable assignment to a string.
+      ContainerOfInts: TypeAlias = "Container[int]"
 
-    Factors: TypeAlias = list[int]
+      class Container(Generic[T]):
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]: Let's name this something else so it doesn't get confused with `typing.Container`. Maybe `Box`?

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -832,19 +832,41 @@ These can be used as types in annotations and do not support ``[]``.
...
+   aliases that make use of forward references, as it can be hard for type
+   checkers to distinguish these from normal variable assignments:
+
+   .. testcode::
+
+      from typing import Generic, TypeAlias, TypeVar
+
+      T = TypeVar("T")
+
+      # "Container" does not exist yet,
```
Comment (AlexWaygood) [inline Doc/library/typing.rst]:
```suggestion
      # "Box" does not exist yet,
```

File: Doc/library/typing.rst (hunk 2 above)
Comment (AlexWaygood) [reply Doc/library/typing.rst]: Good idea, done!

**Comment by JelleZijlstra** [comment]
I'll do the 3.11 backport

**Pull request merged**